File: tests/kicad_xyce_plugin/simulation_parameters/data_block_test.py
from kicad_xyce_plugin.simulation_parameters.data_block import DataBlock


class TestDataBlockParsing:

    def test_parse_single_block_flattened(self):
        # arrange - .DATA block after _join_continuation_lines produces a flat directive
        directives = [
            ".DATA myTable r1 r2 1.0 2.0 3.0 4.0",
            ".ENDDATA",
        ]

        # act
        blocks = DataBlock.from_xyce_directives(directives)

        # assert
        assert len(blocks) == 1
        assert blocks[0].name == "myTable"
        assert blocks[0].parameters == ("r1", "r2")
        assert blocks[0].records == (("1.0", "2.0"), ("3.0", "4.0"))

    def test_parse_single_column_table(self):
        # arrange
        directives = [
            ".DATA temperatures TEMP 25.0 50.0 75.0",
            ".ENDDATA",
        ]

        # act
        blocks = DataBlock.from_xyce_directives(directives)

        # assert
        assert len(blocks) == 1
        assert blocks[0].parameters == ("TEMP",)
        assert blocks[0].records == (("25.0",), ("50.0",), ("75.0",))

    def test_parse_scientific_notation_values(self):
        # arrange
        directives = [
            ".DATA myTable r1 c1 8.0000e+00 4.0000e-09 9.0000e+00 4.0000e-09",
            ".ENDDATA",
        ]

        # act
        blocks = DataBlock.from_xyce_directives(directives)

        # assert
        assert len(blocks) == 1
        assert blocks[0].parameters == ("r1", "c1")
        assert blocks[0].records == (("8.0000e+00", "4.0000e-09"), ("9.0000e+00", "4.0000e-09"))

    def test_parse_multiple_blocks(self):
        # arrange
        directives = [
            ".DATA tableA r1 1.0 2.0",
            ".ENDDATA",
            ".DATA tableB c1 1e-9 2e-9",
            ".ENDDATA",
        ]

        # act
        blocks = DataBlock.from_xyce_directives(directives)

        # assert
        assert len(blocks) == 2
        assert blocks[0].name == "tableA"
        assert blocks[1].name == "tableB"

    def test_no_data_directives_returns_empty(self):
        # arrange
        directives = [".TRAN 1u 1m", ".STEP R1 1k 10k 1k"]

        # act
        blocks = DataBlock.from_xyce_directives(directives)

        # assert
        assert blocks == ()

    def test_spice_suffix_values_recognized(self):
        # arrange - values may use SPICE suffixes like k, meg, n
        directives = [
            ".DATA myTable r1 1k 2k 5k",
            ".ENDDATA",
        ]

        # act
        blocks = DataBlock.from_xyce_directives(directives)

        # assert
        assert len(blocks) == 1
        assert blocks[0].parameters == ("r1",)
        assert blocks[0].records == (("1k",), ("2k",), ("5k",))


class TestDataBlockGeneration:

    def test_generates_data_block_with_header_and_rows(self):
        # arrange
        block = DataBlock(name="myTable", parameters=("r1", "r2"), records=(("1.0", "2.0"), ("3.0", "4.0")))

        # act
        directives = block.to_xyce_directives()

        # assert
        assert directives[0] == ".DATA myTable"
        assert directives[1] == "+ r1 r2"
        assert directives[2] == "+ 1.0 2.0"
        assert directives[3] == "+ 3.0 4.0"
        assert directives[4] == ".ENDDATA"

    def test_generates_single_column_block(self):
        # arrange
        block = DataBlock(name="temps", parameters=("TEMP",), records=(("25.0",), ("75.0",)))

        # act
        directives = block.to_xyce_directives()

        # assert
        assert ".DATA temps" in directives
        assert "+ TEMP" in directives
        assert "+ 25.0" in directives
        assert ".ENDDATA" in directives

    def test_round_trip_preserves_structure(self):
        # arrange
        original = DataBlock(name="sweep", parameters=("r1", "c1"), records=(("1k", "10n"), ("2k", "20n")))

        # act - generate directives then parse them back
        directives = original.to_xyce_directives()
        parsed = DataBlock.from_xyce_directives(directives)

        # assert
        assert len(parsed) == 1
        assert parsed[0].name == original.name
        assert parsed[0].parameters == original.parameters
        assert parsed[0].records == original.records

File: tests/simulation_parameters/data_block.test.cpp
#include <gtest/gtest.h>

#include "simulation_parameters/data_block.h"

// ========================================================================================
// from_xyce_directives
// ========================================================================================

TEST(DataBlockChecks, parse_single_block_flattened) {
    // arrange
    const std::vector<std::string> directives = {
        ".DATA myTable r1 r2 1.0 2.0 3.0 4.0",
        ".ENDDATA",
    };
    // act
    const auto blocks = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(blocks.size(), 1);
    ASSERT_EQ(blocks[0].name, "myTable");
    ASSERT_EQ(blocks[0].parameters.size(), 2);
    ASSERT_EQ(blocks[0].parameters[0], "r1");
    ASSERT_EQ(blocks[0].parameters[1], "r2");
    ASSERT_EQ(blocks[0].records.size(), 2);
    ASSERT_EQ(blocks[0].records[0].size(), 2);
    ASSERT_EQ(blocks[0].records[0][0], "1.0");
    ASSERT_EQ(blocks[0].records[0][1], "2.0");
    ASSERT_EQ(blocks[0].records[1].size(), 2);
    ASSERT_EQ(blocks[0].records[1][0], "3.0");
    ASSERT_EQ(blocks[0].records[1][1], "4.0");
}

TEST(DataBlockChecks, parse_single_column_table) {
    // arrange
    const std::vector<std::string> directives = {
        ".DATA temperatures TEMP 25.0 50.0 75.0",
        ".ENDDATA",
    };
    // act
    const auto blocks = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(blocks.size(), 1);
    ASSERT_EQ(blocks[0].parameters.size(), 1);
    ASSERT_EQ(blocks[0].parameters[0], "TEMP");
    ASSERT_EQ(blocks[0].records.size(), 3);
    ASSERT_EQ(blocks[0].records[0][0], "25.0");
    ASSERT_EQ(blocks[0].records[1][0], "50.0");
    ASSERT_EQ(blocks[0].records[2][0], "75.0");
}

TEST(DataBlockChecks, parse_scientific_notation_values) {
    // arrange
    const std::vector<std::string> directives = {
        ".DATA myTable r1 c1 8.0000e+00 4.0000e-09 9.0000e+00 4.0000e-09",
        ".ENDDATA",
    };
    // act
    const auto blocks = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(blocks.size(), 1);
    ASSERT_EQ(blocks[0].parameters.size(), 2);
    ASSERT_EQ(blocks[0].parameters[0], "r1");
    ASSERT_EQ(blocks[0].parameters[1], "c1");
    ASSERT_EQ(blocks[0].records.size(), 2);
    ASSERT_EQ(blocks[0].records[0][0], "8.0000e+00");
    ASSERT_EQ(blocks[0].records[0][1], "4.0000e-09");
    ASSERT_EQ(blocks[0].records[1][0], "9.0000e+00");
    ASSERT_EQ(blocks[0].records[1][1], "4.0000e-09");
}

TEST(DataBlockChecks, parse_multiple_blocks) {
    // arrange
    const std::vector<std::string> directives = {
        ".DATA tableA r1 1.0 2.0",
        ".ENDDATA",
        ".DATA tableB c1 1e-9 2e-9",
        ".ENDDATA",
    };
    // act
    const auto blocks = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(blocks.size(), 2);
    ASSERT_EQ(blocks[0].name, "tableA");
    ASSERT_EQ(blocks[1].name, "tableB");
}

TEST(DataBlockChecks, no_data_directives_returns_empty) {
    // arrange
    const std::vector<std::string> directives = {
        ".TRAN 1u 1m",
        ".STEP R1 1k 10k 1k",
    };
    // act
    const auto blocks = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(blocks.size(), 0);
}

TEST(DataBlockChecks, spice_suffix_values_recognized) {
    // arrange
    const std::vector<std::string> directives = {
        ".DATA myTable r1 1k 2k 5k",
        ".ENDDATA",
    };
    // act
    const auto blocks = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(blocks.size(), 1);
    ASSERT_EQ(blocks[0].parameters.size(), 1);
    ASSERT_EQ(blocks[0].parameters[0], "r1");
    ASSERT_EQ(blocks[0].records.size(), 3);
    ASSERT_EQ(blocks[0].records[0][0], "1k");
    ASSERT_EQ(blocks[0].records[1][0], "2k");
    ASSERT_EQ(blocks[0].records[2][0], "5k");
}

// ========================================================================================
// to_xyce_directives
// ========================================================================================

TEST(DataBlockChecks, generates_data_block_with_header_and_rows) {
    // arrange
    const DataBlock block("myTable", {"r1", "r2"}, {{"1.0", "2.0"}, {"3.0", "4.0"}});
    // act
    const auto directives = block.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 5);
    ASSERT_EQ(directives[0], ".DATA myTable");
    ASSERT_EQ(directives[1], "+ r1 r2");
    ASSERT_EQ(directives[2], "+ 1.0 2.0");
    ASSERT_EQ(directives[3], "+ 3.0 4.0");
    ASSERT_EQ(directives[4], ".ENDDATA");
}

TEST(DataBlockChecks, generates_data_block_with_single_column) {
    // arrange
    const DataBlock block("temperatures", {"TEMP"}, {{"25.0"}, {"50.0"}, {"75.0"}});
    // act
    const auto directives = block.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 6);
    ASSERT_EQ(directives[0], ".DATA temperatures");
    ASSERT_EQ(directives[1], "+ TEMP");
    ASSERT_EQ(directives[2], "+ 25.0");
    ASSERT_EQ(directives[3], "+ 50.0");
    ASSERT_EQ(directives[4], "+ 75.0");
    ASSERT_EQ(directives[5], ".ENDDATA");
}

// ========================================================================================
// equality operator
// ========================================================================================

TEST(DataBlockChecks, equality_operator_equal_blocks) {
    // arrange
    const DataBlock block1("myTable", {"r1", "r2"}, {{"1.0", "2.0"}});
    const DataBlock block2("myTable", {"r1", "r2"}, {{"1.0", "2.0"}});
    // act
    const bool result = block1 == block2;
    // assert
    ASSERT_TRUE(result);
}

TEST(DataBlockChecks, equality_operator_different_name) {
    // arrange
    const DataBlock block1("myTable", {"r1"}, {{"1.0"}});
    const DataBlock block2("otherTable", {"r1"}, {{"1.0"}});
    // act
    const bool result = block1 == block2;
    // assert
    ASSERT_FALSE(result);
}

TEST(DataBlockChecks, equality_operator_different_parameters) {
    // arrange
    const DataBlock block1("myTable", {"r1"}, {{"1.0"}});
    const DataBlock block2("myTable", {"r2"}, {{"1.0"}});
    // act
    const bool result = block1 == block2;
    // assert
    ASSERT_FALSE(result);
}

TEST(DataBlockChecks, equality_operator_different_records) {
    // arrange
    const DataBlock block1("myTable", {"r1"}, {{"1.0"}});
    const DataBlock block2("myTable", {"r1"}, {{"2.0"}});
    // act
    const bool result = block1 == block2;
    // assert
    ASSERT_FALSE(result);
}

// ========================================================================================
// Additional tests for round-trip
// ========================================================================================

TEST(DataBlockChecks, round_trip_preserves_structure) {
    // arrange
    const DataBlock original("sweep", {"r1", "c1"}, {{"1k", "10n"}, {"2k", "20n"}});
    // act - generate directives then parse them back
    const auto directives = original.to_xyce_directives();
    const auto parsed = DataBlock::from_xyce_directives(directives);
    // assert
    ASSERT_EQ(parsed.size(), 1);
    ASSERT_EQ(parsed[0].name, original.name);
    ASSERT_EQ(parsed[0].parameters.size(), original.parameters.size());
    ASSERT_EQ(parsed[0].parameters[0], original.parameters[0]);
    ASSERT_EQ(parsed[0].parameters[1], original.parameters[1]);
    ASSERT_EQ(parsed[0].records.size(), original.records.size());
    ASSERT_EQ(parsed[0].records[0].size(), original.records[0].size());
    ASSERT_EQ(parsed[0].records[0][0], original.records[0][0]);
    ASSERT_EQ(parsed[0].records[0][1], original.records[0][1]);
    ASSERT_EQ(parsed[0].records[1].size(), original.records[1].size());
    ASSERT_EQ(parsed[0].records[1][0], original.records[1][0]);
    ASSERT_EQ(parsed[0].records[1][1], original.records[1][1]);
}
